### PR TITLE
Bundle CCCL in CuPy

### LINKED ISSUE #6700: Pin CUB upstream version more aggressively?
### Description

After #4897 is concluded CuPy now enables the CUB backend by default. The status quo is
- For CTK < 11.0, CuPy uses the bundled CUB
- For CTK >= 11.0, CuPy uses the CUB from the CUDA Toolkit seen at compile time.

It would be nice if CuPy bundle pins at the CUB (and therefore Thrust) version more aggressively, and does not rely on the CTK. The reason is 2-fold:
- Due to scheduling reasons the CTK's CUB copy usually lags behind the upstream
- Unlike ~2 years ago when we started adding CUB support, now CUB/Thrust are well maintained and constantly having new features/bug fixes added in each release. 

If we lag behind upstream (as the CTK does) we can't enjoy the new additions (ex: supporting #6699 would not be possible as the CUB version needed only comes with CTK 11.6+).

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by leofang**
Close #6700.

~~As per @jrhemstad we expect CCCL v2.2.0 to be tagged soon-ish, but it's not a blocker for this migration 🙂 For now I manually assemble the `cupy/cccl` directory with the three components (thrust/cub/libcudacxx) at v2.1.0, and it'd be easy to update the git submodules later once v2.2.0 is out.~~

*UPDATE*: This PR bundles CCCL (= CUB + Thrust + libcu++) in CuPy. Specifically,
- the bundled CCCL is used at both build- and run- time
- both Jitify and CCCL need to be patched to support NVRTC
  - CCCL patch is merged upstream (NVIDIA/cccl#412) and I updated the submodule to point to the latest CCCL head that includes this patch. This means for now we're a bit advanced than CCCL v2.2.0.
  - Jitify patch to upstream is still work in progress. I managed to find a way to patch it in CuPy, see the updated `cupy/cuda_workaround.h`.
- all header includes are purposely converted to match the CTK layout (ex: `#include <cub/...>`, `#include <thrust/...>`, `#include <cuda/...>`) to avoid issues; also, this ensures that we can pick up from the CuPy bundle, not from CTK, by fixing the header include precedence.
- with the new CCCL layout, many libcu++ headers are picked up for CUB algorithms, and it slows down the first Jitify call to ~50 s on my machine. To mitigate this slowdown, I created a persistent cache similar to the existing kernel cache, see `cupy/cuda/jitify.pyx`. It can be disabled by the new env var `CUPY_DISABLE_JITIFY_CACHE`.

Going forward, all `type_traits` patches in CuPy/Jitify should be converted to include `<cuda/std/type_traits>`. I'll open a new issue to track this. It's not a blocker.

### CODE REVIEW

**Comment by leofang** [comment]
TODO: Modify include paths fed to NVRTC at run time.

**Comment by leofang** [comment]
Working on a CUB patch so that it's NVRTC-compilable: https://github.com/NVIDIA/cccl/pull/412

**Comment by leofang** [comment]
@kmaehashi Am I actually waiting in the CI queue, or somehow I don't have permission to launch all jobs? 🤔

**Comment by kmaehashi** [comment]
:eyes: https://github.com/cupy/cupy/actions/runs/6123693803/job/16622120429#step:6:1

To launch all tests, you need to say:
/test full

**Comment by leofang** [comment]
ahhh silly me... thanks!

**Comment by leofang** [comment]
I am clueless why the Windows CIs would fail at build time...
```
copying cupy\_core\include\cupy\cccl\libcudacxx\libcxx\test\std\experimental\memory\memory.polymorphic.allocator.class\memory.polymorphic.allocator.mem\construct_pair_const_lvalue_pair.pass.cpp -> build\lib.win-amd64-cpython-39\cupy\_core\include\cupy\cccl\libcudacxx\libcxx\test\std\experimental\memory\memory.polymorphic.allocator.class\memory.polymorphic.allocator.mem
error: could not create 'build\lib.win-amd64-cpython-39\cupy\_core\include\cupy\cccl\libcudacxx\libcxx\test\std\experimental\memory\memory.polymorphic.allocator.class\memory.polymorphic.allocator.mem\construct_pair_const_lvalue_pair.pass.cpp': No such file or directory
```

**Comment by kmaehashi** [comment]
It's copying a file from:
```
C:\Windows\TEMP\flexci\run-00292523\work\src\cupy\_core\include\cupy\cccl\libcudacxx\libcxx\test\std\experimental\memory\memory.polymorphic.allocator.class\memory.polymorphic.allocator.mem\construct_pair_const_lvalue_pair.pass.cpp
```
to:
```
C:\Windows\TEMP\flexci\run-00292523\work\src\build\lib.win-amd64-cpython-39\cupy\_core\include\cupy\cccl\libcudacxx\libcxx\test\std\experimental\memory\memory.polymorphic.allocator.class\memory.polymorphic.allocator.mem\construct_pair_const_lvalue_pair.pass.cpp
```
and the latter exceeds Windows's MAX_PATH (260) chars. I'll try updating the registry key to workaroud the issue.

**Comment by leofang** [comment]
(so the tags are taken as a union, not an intersection 😂)

**Comment by leofang** [comment]
> Working on a CUB patch so that it's NVRTC-compilable: [NVIDIA/cccl#412](https://github.com/NVIDIA/cccl/pull/412)

Update: this is merged and I updated the submodule to point to the latest CCCL head that includes this patch. A question in discussion internally is whether we'd release CCCL v2.2.x, if so I'd ask for backporting this patch.

**Comment by leofang** [comment]
@kmaehashi what's the right way of addressing CI timeout? Just retry until the cache is populated?

**Comment by kmaehashi** [comment]
The failing CI `cupy.win.cuda112` seems to be a real error rather than timeout?

```
00:10:42.750065 STDERR 1048]	  c:\windows\temp\flexci\run-00293065\work\src\cupy\_core\include\cupy\cccl\libcudacxx\include\cuda\std\detail\libcxx\include\__utility/unreachable.h(28): error: identifier "__builtin_unreachable" is undefined	
```

**Comment by leofang** [comment]
Yes that one I'd have to address later, don't know what happens, but I think the rest is working.

**Comment by kmaehashi** [comment]
Now that #7775 got merged, the cache started to grow. It should be fully cashed after some more runs in the main branch.

**Comment by leofang** [comment]
> It should be fully cached after some more runs in the main branch.

@kmaehashi sorry I may have misunderstood: Do you mean that the cache would only be populated after merged into main, not in PRs? If so I should stop retrying and instead focus on fixing Win + cuda112.

**Comment by leofang** [comment]
All CI (except for win.cuda112) failed due to timeout.

**Comment by leofang** [comment]
> win.cuda112 has quite some test failures, but because the CI timed out, the error log was truncated and I could not see the detail, not sure what to do... 😅

It should be hopefully fixed by commit ecb6270, which I planned to do in #7869. It seems we got #4383 fixed for free by ensuring that NVRTC sees a consistent set of CCCL headers.

**Comment by leofang** [comment]
This is ready!

**Comment by kmaehashi** [comment]
> Do you mean that the cache would only be populated after merged into main, not in PRs?

Sorry I dropped the ball. Yes, the cache is updated only in the branch CI to avoid pollution.

**Comment by leofang** [comment]
I see. Is there a way we can temporarily increase the timeout limit? I'd prefer to see all CIs green before merging (though if not possible I can always send follow-up PRs to fix them) 🙂

**Comment by kmaehashi** [comment]
Could you try extending the timeout [here](https://github.com/cupy/cupy/blob/5c32e40af32f6f9627e09d47ecfeb7e9281ccab2/.pfnci/config.pbtxt#L16) (the overall timeout - including cache download + build + pytest) and [here](https://github.com/cupy/cupy/blob/5c32e40af32f6f9627e09d47ecfeb7e9281ccab2/.pfnci/linux/tests/actions/unittest.sh#L26) (pytest timeout - 18000 sec)?

**Comment by leofang** [comment]
it seems we still hit timeout on windows?

**Comment by leofang** [comment]
@kmaehashi any suggestion? 🙂

**Comment by kmaehashi** [comment]
```
00:20:12.284651 STDERR 2628]	RunWithTimeout : Cannot validate argument on parameter 'Timeout'. The 36000	
00:20:12.284651 STDERR 2628]	argument is greater than the maximum allowed range of 32767. Supply an	
00:20:12.284651 STDERR 2628]	argument that is less than or equal to 32767 and then try the command again.	
```

So the maximum timeout on Windows seems 32767 seconds... 😞

File: cupy/_core/core.pyx [hunk 1]
```diff
@@ -2197,8 +2204,22 @@ cpdef function.Module compile_with_cache(
 
     if prepend_cupy_headers:
         source = _cupy_header + source
+    if jitify:
+        source = '#include <cupy/cuda_workaround.h>\n' + source
     extra_source = _get_header_source()
-    options += ('-I%s' % _get_header_dir_path(),)
+
+    for op in options:
+        if '-std' in op:
```
Comment (kmaehashi) [inline cupy/_core/core.pyx:2212]: This looks a bit fragile 😅 
User specifying `-I/foo/bar-std/baz` sounds likely.

Comment (leofang) [reply cupy/_core/core.pyx:2212]: Yeah, I think checking `-std=` (with an extra character `=`) should be more robust?

Comment (leofang) [reply cupy/_core/core.pyx:2212]: or `-std=c++`...

Comment (kmaehashi) [reply cupy/_core/core.pyx:2212]: > `-std=c++`

This sounds good to me!
(Strictly speaking `('-std', 'c++11')` is also a valid option but I guess no one do this)

Comment (leofang) [reply cupy/_core/core.pyx]: Done with [e381c2a](https://github.com/cupy/cupy/pull/7851/commits/e381c2a780573e731090d9d76c19ad0dda5dec52).

**Comment by leofang** [comment]
> So the maximum timeout on Windows seems 32767 seconds... 😞

@kmaehashi so should I revert or modify commit https://github.com/cupy/cupy/pull/7851/commits/fd00150534c0aef3f803487c7a09262d31dcc124?

**Comment by kmaehashi** [comment]
Could you try replacing `36000` with `32767` (9.1 hours) in this line?

```
    $test_retval = RunWithTimeout -timeout 36000 -output ../cupy_test_log.txt -- python -m pytest -rfEX @pytest_opts .
```

If this attempt fails, probably we need to remove the timeout...

**Comment by leofang** [comment]
@kmaehashi it seems windows is all green (including cuda122!)

**Review by kmaehashi — APPROVED** [review]

**Pull request merged**